Support for CloudStack and thus Ninefold's compute platform

jeamland opened the pull request:
These commits add support for cloud.com's CloudStack API and thus allow the use of Ninefold (ninefold.com) as a compute/loadbalancer provider.

jeamland committed:
Add beginnings of Ninefold compute driver.

jeamland committed:
Add more functions for listing various available things.

jeamland committed:
Refactor _api_request.

jeamland committed:
Reformat NODE_STATE_MAP, add name.

jeamland committed:
Add reboot and destroy methods.

jeamland committed:
Refactor request logic.

jeamland committed:
Add more information in list_nodes.

jeamland committed:
Add create_node.

jeamland committed:
Add extension methods for handling public addresses and forwarding rules.

jeamland committed:
Fix typo.

jeamland committed:
Merge branch 'trunk' of https://github.com/apache/libcloud into trunk

jeamland committed:
Rework a few things to better support making this a generic CloudStack driver.

jeamland committed:
Docstrings.

jeamland committed:
Move _sync_request and _async_request to the connection.

jeamland committed:
Turn the Ninefold driver into a generic CloudStack driver.

jeamland committed:
Move CloudStack connection handling stuff into libcloud.common.

jeamland committed:
Get MalformedResponseError from a better place.

jeamland committed:
Move the signature algorithm into its own method.

jeamland committed:
Fix error handling in sync requests.

jeamland committed:
Fix async polling.

jeamland committed:
Add tests for CloudStack connection handling.

jeamland committed:
Fix a value lookup.

jeamland committed:
Add basic tests for CloudStack.

jeamland committed:
Add another signature test.

jeamland committed:
Merge branch 'trunk' of https://github.com/apache/libcloud into trunk

jeamland committed:
Add test for create_node failure.

jeamland committed:
Whoops, DeploymentError pertains to deploy_node, not create_node.

jeamland committed:
Test for delayed failure as well.

jeamland committed:
Add host and port to this to bring it in line with the compute equivalent.

jeamland committed:
Move some logic from the node driver to a mix-in so we can use it for the load balancer code.

jeamland committed:
Initial load balancer support.

jeamland committed:
Missed test fixture.

jeamland committed:
Add create/destroy.

jeamland committed:
Switch _async_request around so it behaves a bit more like _sync_request.

jeamland committed:
Tests for CloudStack load balancer.

jeamland committed:
Handle having no load balancers.

jeamland committed:
Add a couple of docstrings.

jeamland committed:
Add test path for common.

jeamland committed:
Merge branch 'trunk' of https://github.com/apache/libcloud into trunk

Kami commented on libcloud/common/cloudstack.py:
```diff
@@ -0,0 +1,101 @@
+        return params, headers
+
+    def _sync_request(self, command, **kwargs):
+        "Make a synchronous API request. These return immediately."
```
Hm, you have probably meant "These **don't** return immediately.", because a synchronous request blocks :)

jeamland replied:
Commit af73bfe addresses this. Thanks!

Kami commented on libcloud/common/cloudstack.py:
```diff
@@ -0,0 +1,101 @@
+            time.sleep(self.driver.async_poll_frequency)
+
+        if result['jobstatus'] == 2:
+            raise Exception(result)
```
Hm, what does jobstatus 2 means? We should probably throw a more granular exception.

jeamland replied:
Commit 13b7fbf addresses this.  The job can either be still pending, have succeeded or have failed.  Thanks.

Kami commented on libcloud/compute/drivers/cloudstack.py:
```diff
@@ -0,0 +1,250 @@
+            id=node['id'],
+            name=node['displayname'],
+            state=self.NODE_STATE_MAP[node['state']],
+            public_ip=[],
```
Hm, by default Node has no public IP address assigned? We should probably request a public ip address here as well so it's consistent with other drivers.

Also, if Node has no public IP address, `deploy_node` can't be used.

jeamland replied:
As discussed, Ninefold would rather the act of assigning a public IP be explicit.

Maybe NodeDriver could grow a feature to indicate whether a public IP is assigned by default and the deploy_node could call a hook to allocate the IP if that feature wasn't present?

Kami commented on libcloud/compute/drivers/cloudstack.py:
```diff
@@ -0,0 +1,250 @@
+        szs = self._sync_request('listServiceOfferings')
+        sizes = []
+        for sz in szs['serviceoffering']:
+            sizes.append(NodeSize(sz['id'], sz['name'], sz['memory'], 0, 0,
```
So Ninefold doesn't exposes prices through the API. Do they have them listed on the website? If they do, you should put them in `data/pricing.json`...

jeamland replied:
I can sort that out.  What units is the pricing in though? Ninefold charge in AUD rather than USD...

Kami replied:
Currently we just convert all the prices to USD which is obviously a bad idea.

I was thinking about adding a new `Price` class which would also have a `currency` attribute.

jeamland replied:
Ok, it's even more complicated than that actually.  Ninefold's pricing consists of a component for the instance size and then another for the template image, both per hour.  The template image cost varies depending on the hypervisor you're using (they support XenServer as well as VMware) and the OS (Linux flavours or Windows).

I'm really not sure if I can fit that into your current pricing model.

Kami replied:
Yeah, it's probably not worth the hassle. At least for the first version.

Kami commented:
Hm, I  have clone your repository and two tests fail:

```
...
======================================================================
ERROR: test_async_request_successful (test.common.test_cloudstack.CloudStackCommonTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kami/Projects/libcloud/libcloud-ninefold/test/common/test_cloudstack.py", line 49, in test_async_request_successful
    success, result = self.connection._async_request('fake')
ValueError: need more than 1 value to unpack

======================================================================
ERROR: test_async_request_unsuccessful (test.common.test_cloudstack.CloudStackCommonTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kami/Projects/libcloud/libcloud-ninefold/test/common/test_cloudstack.py", line 55, in test_async_request_unsuccessful
    success, _ = self.connection._async_request('fake')
  File "/Users/kami/Projects/libcloud/libcloud-ninefold/libcloud/common/cloudstack.py", line 81, in _async_request
    raise Exception(result)
Exception: {u'jobstatus': 2, u'jobresult': {u'fake': u'failresult'}}

----------------------------------------------------------------------
Ran 684 tests in 5.109s

FAILED (errors=2)
```

Kami commented on libcloud/loadbalancer/drivers/cloudstack.py:
```diff
@@ -0,0 +1,106 @@
+            raise Exception("no such load balancer: " + str(balancer_id))
+        return self._to_balancer(balancer[0])
+
+    def create_balancer(self, name, members, protocol='http', port=80,
```
Hm, yeah, location should probably be part of the base API, similar to the base compute API...

Will ask what other ppl think about it.

jeamland committed:
Fix a couple of tests.

jeamland committed:
Make docstrings on _sync_request and _async_request more clear.

jeamland committed:
Use named constants rather than magic numbers for async job status.

jeamland committed:
Merge branch 'trunk' of https://github.com/apache/libcloud into trunk

jeamland committed:
Add license.

jeamland committed:
Fix names and api_names.

jeamland committed:
Merge branch 'trunk' of https://github.com/apache/libcloud into trunk

jeamland committed:
Merge remote-tracking branch 'apache/trunk' into trunk

Conflicts:
	libcloud/common/cloudstack.py
	libcloud/compute/drivers/cloudstack.py
	libcloud/compute/providers.py
	libcloud/compute/types.py
	libcloud/loadbalancer/base.py
	libcloud/loadbalancer/drivers/cloudstack.py
	libcloud/loadbalancer/providers.py
	libcloud/loadbalancer/types.py
	setup.py

jeamland committed:
Fix escaping of names in the Atmos driver.

Hat tip to @freakboy3742

jeamland committed:
Pass kwargs from container.download_object to driver.download_object.

jeamland committed:
Merge remote-tracking branch 'apache/trunk' into trunk

jeamland committed:
Fix upload_object_via_stream in the Atmos driver.

The existing code would only work in the case where the object already
existed in the container. This version handles creating the object if
needed.

jeamland committed:
Add tests to cover both creation and replacement in upload_object_via_stream.

freakboy3742 committed:
Ensure that content type is preserved during streaming upload to atmos.

jeamland committed:
Merge pull request #1 from freakboy3742/trunk

Ensure that content type is preserved during streaming upload to atmos.

Kami commented:
(just doing some pull request cleanup)

@jeamland - Those changes have been merged a long time ago, can you please close this pull request?

Thanks!
